### ISSUE TITLE
Pod stuck in terminating when subPath was deleted in azure fileshare

### ISSUE THREAD

**Issue by Jiawei0227** [kubernetes/kubernetes#97031]
**What happened**:
I was running the subpath testsuite against Azurefile.
This test failed and can be easily reproduced:
`should be able to unmount after the subpath directory is deleted [LinuxOnly]`

Basically what this test does is:

1. Create a PVC use azurefile
2. Create a pod that has two containers. One of it mounts the volume itself. The second container mounts the subpath of the volume.
3. Delete the subpath dir from the container that has the whole volume
4. Delete the pod, the pod will stuck in terminating status and cannot be cleaned up other than `--force`

**What you expected to happen**:
The pod can be deleted successfully.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create the pvc
```yaml
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-volume-pvc
spec:
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  storageClassName: azurefile-csi
```

2. Create the pod
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod-azurefile
  labels:
    app: myapp
spec:
  containers:
  - name: myapp-container-parent
    image: docker.io/library/busybox:1.29
    securityContext:
      privileged: true
    command: ['sh', '-c', 'while true; do echo The app is running! && sleep 3600; done']
    volumeMounts:
    - name: test-volume
      mountPath: /test-volume
  - name: myapp-container-subpath
    image: docker.io/library/busybox:1.29
    securityContext:
      privileged: true
    command: ['sh', '-c', 'while true; do echo The app is running! && sleep 3600; done']
    volumeMounts:
    - name: test-volume
      mountPath: /test-volume
      subPath: provisioning
  volumes:
  - name: test-volume
    persistentVolumeClaim:
      claimName: test-volume-pvc
```

3. delete the subpath dir from the parent pod
```
$ kubectl exec -it myapp-pod-azurefile --container myapp-container-parent -- /bin/bash rm -r /test-volume/provisioning
```

4. delete the pod
```
$ kubectl delete pod myapp-pod-azurefile
```

**Anything else we need to know?**:
The volume is being formatted as cifs type

**Environment**:
- Kubernetes version (use `kubectl version`): v1.18.6
- Cloud provider or hardware configuration: Azurefile-csi-driver 0.9.0
- OS (e.g: `cat /etc/os-release`): Ubuntu 20.04
- Kernel (e.g. `uname -a`): Linux nodepool-f985db88-68ff3c05000000 5.4.0-1016-azure #16-Ubuntu SMP Thu Jun 4 22:49:15 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by Jiawei0227** [comment]
1. kubelet log
```
Dec 02 02:48:48 nodepool-f985db88-68ff3c05000000 kubelet[3290]: I1202 02:48:48.209041    3290 reconciler.go:196] operationExecutor.UnmountVolume started for volume "test-volume" (UniqueName: "kubernetes.io/csi/file.csi.azure.com^anthos_cluster_f985db88#f0ded069e835745
88#f0ded069e835745cbac4ed7#pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca#") pod "f5d2e3b4-dc8e-42bc-a35a-59428db817fd" (UID: "f5d2e3b4-dc8e-42bc-a35a-59428db817fd")
Dec 02 02:48:48 nodepool-f985db88-68ff3c05000000 kubelet[3290]: W1202 02:48:48.221645    3290 mount_helper_common.go:33] Warning: Unmount skipped because path does not exist: /var/lib/kubelet/pods/f5d2e3b4-dc8e-42bc-a35a-59428db817fd/volume-subpaths/pvc-8feaa985-0907-
pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca/myapp-container/0
Dec 02 02:48:48 nodepool-f985db88-68ff3c05000000 kubelet[3290]: E1202 02:48:48.222127    3290 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/file.csi.azure.com^anthos_cluster_f985db88#f0ded069e835745cbac4ed7#pvc-8feaa985-0907-422e-b3a9-b2
-0907-422e-b3a9-b2e18450e6ca# podName:f5d2e3b4-dc8e-42bc-a35a-59428db817fd nodeName:}" failed. No retries permitted until 2020-12-02 02:50:50.222062631 +0000 UTC m=+983128.678263814 (durationBeforeRetry 2m2s). Error: "error cleaning subPath mounts fovome\" (UniqueName
r volume \"test-volume\" (UniqueName: \"kubernetes.io/csi/file.csi.azure.com^anthos_cluster_f985db88#f0ded069e835745cbac4ed7#pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca#\") pod \"f5d2e3b4-dc8e-42bc-a35a-59428db817fd\" (UID: \"f5d2e3b4-dc8e-42bc-a35a-59481/kubelet/pods/f5
28db817fd\") : error deleting /var/lib/kubelet/pods/f5d2e3b4-dc8e-42bc-a35a-59428db817fd/volume-subpaths/pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca/myapp-container: remove /var/lib/kubelet/pods/f5d2e3b4-dc8e-42bc-a35a-59428db817fd/volume-subpaths/pvc-85-ectory not empty
feaa985-0907-422e-b3a9-b2e18450e6ca/myapp-container: directory not empty"
```

After debugging with @msau42 and @jingxu97 offline, the issue is that:

After the dir from the subpath was removed, the subpath mountpoint on the host machine becomes into a weird state where you can still find the dir by doing a readdir to its parent. But doing os.stat or something like that will result in a "No such file or directory".

When you do a ls on the path:
```
root@nodepool-f985db88-68ff3c05000000:/home/gkeadmin/oscheck# ll /var/lib/kubelet/pods/f1f99fc4-d5a4-487c-a76a-f522b088feab/volume-subpaths/pvc-741b65ae-487a-48a4-9e90-1c972b778d69/myapp-container
ls: cannot access '/var/lib/kubelet/pods/f1f99fc4-d5a4-487c-a76a-f522b088feab/volume-subpaths/pvc-741b65ae-487a-48a4-9e90-1c972b778d69/myapp-container/0': No such file or directory
total 8
drwxr-x--- 3 root root 4096 Dec  2 21:55 ./
drwxr-x--- 3 root root 4096 Dec  2 21:55 ../
d????????? ? ?    ?       ?            ? 0/
```

**Comment by msau42** [comment]
@kubernetes/sig-storage-bugs

**Comment by Jiawei0227** [comment]
So the issue from code is that 
https://github.com/kubernetes/kubernetes/blob/5b8c3b90f3c74ba79f4bfa5be560d011694a7da0/staging/src/k8s.io/mount-utils/mount_helper_common.go#L33

It will fail the path exists check while the actual mountpoint exists. So when the code logic runs to here:
https://github.com/kubernetes/kubernetes/blob/978233775e742fcd1f810ced72c9948bc67427cc/pkg/volume/util/subpath/subpath_linux.go#L269

It will fail to remove the container dir because it somehow detect there is still file there.

To fix this as a workaround, user can manually `umount /var/lib/kubelet/pods/xxx/path/to/subpath/volume`, and then the kubelet can take care of this. 

To fix it ultimately, we need to see if this is an issue specific to cifs filesystem? Or we should consider do an umount regardless if the path exists.

**Comment by nearora-msft** [comment]
> So the issue from code is that
> https://github.com/kubernetes/kubernetes/blob/5b8c3b90f3c74ba79f4bfa5be560d011694a7da0/staging/src/k8s.io/mount-utils/mount_helper_common.go#L33
> 
> It will fail the path exists check while the actual mountpoint exists. So when the code logic runs to here:
> https://github.com/kubernetes/kubernetes/blob/978233775e742fcd1f810ced72c9948bc67427cc/pkg/volume/util/subpath/subpath_linux.go#L269
> 
> It will fail to remove the container dir because it somehow detect there is still file there.
> 
> To fix this as a workaround, user can manually `umount /var/lib/kubelet/pods/xxx/path/to/subpath/volume`, and then the kubelet can take care of this.
> 
> To fix it ultimately, we need to see if this is an issue specific to cifs filesystem? Or we should consider do an umount regardless if the path exists.

Tried the same scenario using NFS. It works as expected using NFS.

**Comment by Jiawei0227** [comment]
> Tried the same scenario using NFS. It works as expected using NFS.

Yeah, that's why I am wondering if this is something special with CIFS?

**Comment by nearora-msft** [comment]
> > Tried the same scenario using NFS. It works as expected using NFS.
> 
> Yeah, that's why I am wondering if this is something special with CIFS?

Looks like it has something to do with the  azurefile-csi storage class mount options.
https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/c6e5eb17b3a875fef30ebc6871e32f11fa061297/deploy/example/pv-azurefile-csi.yaml#L13

I was getting this error when I'm trying to do ls on the subpath. So I realized there might be some issue with the file permissions.

root@aks-ultradisk-21605368-vmss000001:/var/lib/kubelet/pods/8775fedf-da46-4457-b8eb-a1ab2d71a816/volume-subpaths/pvc-9abe3530-e16d-48f5-b0cd-a44ae94d7d55/myapp-container-subpath/0# ls
ls: cannot open directory '.': No such file or directory

When I removed the dir_mode and file_mode properties from the storageclass and performed the same steps as mentioned in the issue, the pod got deleted successfully.(Although it took a lot of time 10-15 mins for the pod to get deleted)

**Comment by andyzhangx** [comment]
> > Tried the same scenario using NFS. It works as expected using NFS.
> 
> Yeah, that's why I am wondering if this is something special with CIFS?

@Jiawei0227 FYI, I am planning to use this PR to fix the unmount issue in Azure File & SMB CSI driver: https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/671
Not sure how to handle gracefully in k8s subPath code since that's a common code.

**Comment by Jiawei0227** [comment]
> > > Tried the same scenario using NFS. It works as expected using NFS.
> > 
> > 
> > Yeah, that's why I am wondering if this is something special with CIFS?
> 
> @Jiawei0227 FYI, I am planning to use this PR to fix the unmount issue in Azure File & SMB CSI driver: [kubernetes-sigs/azurefile-csi-driver#671](https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/671)
> Not sure how to handle gracefully in k8s subPath code since that's common code.

So this PR does not fix the Azurefile subpath test issue?

**Comment by andyzhangx** [comment]
> > > > Tried the same scenario using NFS. It works as expected using NFS.
> > > 
> > > 
> > > Yeah, that's why I am wondering if this is something special with CIFS?
> > 
> > 
> > @Jiawei0227 FYI, I am planning to use this PR to fix the unmount issue in Azure File & SMB CSI driver: [kubernetes-sigs/azurefile-csi-driver#671](https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/671)
> > Not sure how to handle gracefully in k8s subPath code since that's common code.
> 
> So this PR does not fix the Azurefile subpath test issue?

@Jiawei0227 no, that PR only fixed the unmount issue inside CSI driver. While for the subPath issue, we should fix it in k8s upstream code as you mentioned above: https://github.com/kubernetes/kubernetes/issues/97031#issuecomment-737636695

**Comment by andyzhangx** [comment]
> 1. kubelet log
> 
> ```
> Dec 02 02:48:48 nodepool-f985db88-68ff3c05000000 kubelet[3290]: I1202 02:48:48.209041    3290 reconciler.go:196] operationExecutor.UnmountVolume started for volume "test-volume" (UniqueName: "kubernetes.io/csi/file.csi.azure.com^anthos_cluster_f985db88#f0ded069e835745
> 88#f0ded069e835745cbac4ed7#pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca#") pod "f5d2e3b4-dc8e-42bc-a35a-59428db817fd" (UID: "f5d2e3b4-dc8e-42bc-a35a-59428db817fd")
> Dec 02 02:48:48 nodepool-f985db88-68ff3c05000000 kubelet[3290]: W1202 02:48:48.221645    3290 mount_helper_common.go:33] Warning: Unmount skipped because path does not exist: /var/lib/kubelet/pods/f5d2e3b4-dc8e-42bc-a35a-59428db817fd/volume-subpaths/pvc-8feaa985-0907-
> pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca/myapp-container/0
> Dec 02 02:48:48 nodepool-f985db88-68ff3c05000000 kubelet[3290]: E1202 02:48:48.222127    3290 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/file.csi.azure.com^anthos_cluster_f985db88#f0ded069e835745cbac4ed7#pvc-8feaa985-0907-422e-b3a9-b2
> -0907-422e-b3a9-b2e18450e6ca# podName:f5d2e3b4-dc8e-42bc-a35a-59428db817fd nodeName:}" failed. No retries permitted until 2020-12-02 02:50:50.222062631 +0000 UTC m=+983128.678263814 (durationBeforeRetry 2m2s). Error: "error cleaning subPath mounts fovome\" (UniqueName
> r volume \"test-volume\" (UniqueName: \"kubernetes.io/csi/file.csi.azure.com^anthos_cluster_f985db88#f0ded069e835745cbac4ed7#pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca#\") pod \"f5d2e3b4-dc8e-42bc-a35a-59428db817fd\" (UID: \"f5d2e3b4-dc8e-42bc-a35a-59481/kubelet/pods/f5
> 28db817fd\") : error deleting /var/lib/kubelet/pods/f5d2e3b4-dc8e-42bc-a35a-59428db817fd/volume-subpaths/pvc-8feaa985-0907-422e-b3a9-b2e18450e6ca/myapp-container: remove /var/lib/kubelet/pods/f5d2e3b4-dc8e-42bc-a35a-59428db817fd/volume-subpaths/pvc-85-ectory not empty
> feaa985-0907-422e-b3a9-b2e18450e6ca/myapp-container: directory not empty"
> ```
> 
> After debugging with @msau42 and @jingxu97 offline, the issue is that:
> 
> After the dir from the subpath was removed, the subpath mountpoint on the host machine becomes into a weird state where you can still find the dir by doing a readdir to its parent. But doing os.stat or something like that will result in a "No such file or directory".
> 
> When you do a ls on the path:
> 
> ```
> root@nodepool-f985db88-68ff3c05000000:/home/gkeadmin/oscheck# ll /var/lib/kubelet/pods/f1f99fc4-d5a4-487c-a76a-f522b088feab/volume-subpaths/pvc-741b65ae-487a-48a4-9e90-1c972b778d69/myapp-container
> ls: cannot access '/var/lib/kubelet/pods/f1f99fc4-d5a4-487c-a76a-f522b088feab/volume-subpaths/pvc-741b65ae-487a-48a4-9e90-1c972b778d69/myapp-container/0': No such file or directory
> total 8
> drwxr-x--- 3 root root 4096 Dec  2 21:55 ./
> drwxr-x--- 3 root root 4096 Dec  2 21:55 ../
> d????????? ? ?    ?       ?            ? 0/
> ```

@Jiawei0227 @msau42 @jingxu97 back to original discussion, shall we use `os.Lstat` instead of `os.Stat` func? That will solve the issue of smb file PathExists check issue.

https://github.com/kubernetes/kubernetes/blob/5b8c3b90f3c74ba79f4bfa5be560d011694a7da0/staging/src/k8s.io/mount-utils/mount_helper_common.go#L91-L102

```
// Stat returns a FileInfo describing the named file.
// If there is an error, it will be of type *PathError.
func Stat(name string) (FileInfo, error) {
	testlog.Stat(name)
	return statNolog(name)
}

// Lstat returns a FileInfo describing the named file.
// If the file is a symbolic link, the returned FileInfo
// describes the symbolic link. Lstat makes no attempt to follow the link.
// If there is an error, it will be of type *PathError.
func Lstat(name string) (FileInfo, error) {
	testlog.Stat(name)
	return lstatNolog(name)
}
```

**Comment by andyzhangx** [comment]
this subPath issue was fixed in `1.21.9`, `1.22.6`, `1.23.2`
